File: lab7-vars/haskell/accomRoomSize.hs
{-# OPTIONS_GHC -fno-warn-missing-signatures -fno-warn-incomplete-patterns #-}
module Main where

data RoomCategory =
    Launge | Bedroom | Kitchen | Bathroom | DiningRoom | UtilityRoom | WC
    deriving (Show, Eq) -- default formatting and equality comparison

data Room =
    Room
    {
        roomCategory :: RoomCategory,
        roomSizeInCM :: (Int, Int)
    }

instance Show Room where -- how to format values of type Room
    show (Room cat (x,y)) =
        show cat ++ " " ++ show x ++ "cm x " ++ show y ++ "cm"

roomOK (Room _ (x,y)) = x > 0 && y > 0

roomArea room =
    x * y
    where
    x = fst (roomSizeInCM room)
    y = snd (roomSizeInCM room)

totalArea rooms =
    sum [roomArea room | room <- rooms, roomOK room]

myRooms =
    [Room Kitchen (450,310), Room Launge (550,450),
     Room Bathroom (250,180), Room Bedroom (450,390)]

main =
    do
    putStr $ unlines [show room | room <- myRooms]
    putStrLn $ "total area = " ++ (show (totalArea myRooms)) ++ "cm"

File: lab7-vars/java/accomRoomSize.java
import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.HashMap;

public class accomRoomSize
{

    private enum RoomKind { Launge, Bedroom, Kitchen, Bathroom, DiningRoom, UtilityRoom, WC }
    
    private static class Room
    {
        public Room(RoomKind kind, int sizeXinCm, int sizeYinCm)
        {
            this.kind = kind;
            this.sizeXinCm = sizeXinCm;
            this.sizeYinCm = sizeYinCm;
        }

        public RoomKind getKind(){ return kind; }
        public int getSizeXinCm(){ return sizeXinCm; }
        public int getSizeYinCm(){ return sizeYinCm; }

        public String toString()
        {
            return "" + kind + " " + sizeXinCm + "cm x " + sizeYinCm + "cm";
        }

        private RoomKind kind;
        private int sizeXinCm;
        private int sizeYinCm;
    }

    private static <T> void printListPerLine(String prompt, List<T> list)
    {
        System.out.println(prompt);
        for(T item : list)
        {
            System.out.println(" " + item);
        }
    }

    private static int totalArea(List<Room> rooms)
    {
        int result = 0;
        for(Room room : rooms)
        {
            int x = room.getSizeXinCm();
            int y = room.getSizeYinCm();
            result += x * y;
        }
        return result;
    }

    public static void main(String[] args)
    {
        List<Room> myrooms = new ArrayList<Room>();
        myrooms.add(new Room(RoomKind.Kitchen, 450, 310));
        myrooms.add(new Room(RoomKind.Launge, 550, 450));
        myrooms.add(new Room(RoomKind.Bathroom, 250, 180));

        printListPerLine("my rooms = ", myrooms);
        System.out.println("total area = " + totalArea(myrooms) + "cm");

        myrooms.add(new Room(RoomKind.Bedroom, 450, 390));
        printListPerLine("my rooms = ", myrooms);
        System.out.println("total area = " + totalArea(myrooms) + "cm");
   }
}
